packages/kbn-generate-csv/src/generate_csv_esql.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { lastValueFrom } from 'rxjs';
+import type { Writable } from 'stream';
+
+import { errors as esErrors } from '@elastic/elasticsearch';
+import type { IScopedClusterClient, IUiSettingsClient, Logger } from '@kbn/core/server';
+import type { ESQLSearchParams, ESQLSearchReponse } from '@kbn/es-types';
+import {
+  cellHasFormulas,
+  ESQL_SEARCH_STRATEGY,
+  getEsQueryConfig,
+  IKibanaSearchRequest,
+  IKibanaSearchResponse,
+} from '@kbn/data-plugin/common';
+import type { IScopedSearchClient } from '@kbn/data-plugin/server';
+import {
+  AuthenticationExpiredError,
+  byteSizeValueToNumber,
+  CancellationToken,
+  ReportingError,
+} from '@kbn/reporting-common';
+import type { TaskRunResult } from '@kbn/reporting-common/types';
+import type { ReportingConfigType } from '@kbn/reporting-server';
+import { buildEsQuery, Filter } from '@kbn/es-query';
+import { zipObject } from 'lodash';
+
+import { CONTENT_TYPE_CSV } from './constants';
+import { CsvExportSettings, getExportSettings } from './get_export_settings';
+import { i18nTexts } from './i18n_texts';
+import { MaxSizeStringBuilder } from './max_size_string_builder';
+
+export interface JobParamsCsvESQL {
+  query: { esql: string };
+  columns?: string[];
+  filters?: Filter[];
+  browserTimezone?: string;
+}
+
+interface Clients {
+  es: IScopedClusterClient;
+  data: IScopedSearchClient;
+  uiSettings: IUiSettingsClient;
+}
+
+export class CsvESQLGenerator {
```

Comment:
This is a copy of `CsvGenerator` with changes for es|ql generation. I kept most of the CSV building-specific things like maxSizeBytes, bom, escapeFormulaValues, cancelation, and CPU breaks, and reused the same CSV settings. Inputs/Outputs/errors/warnings are mostly the same